vg/recorder/recorder.go

```diff
@@ -0,0 +1,458 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package recorder provides support for vector graphics serialization
+// it is intended to be used as a testing helper.
+package recorder
+
+import (
+	"fmt"
+	"image/color"
+	"runtime"
+
+	"github.com/gonum/plot/vg"
+)
+
+var _ vg.Canvas = (*Canvas)(nil)
+
+// Canvas implements vg.Canvas operation serialization.
+type Canvas struct {
+	// Resolution holds the canvas resolution in DPI.
+	Resolution float64
+
+	// Actions holds the complete state of the canvas.
+	Actions []Action
+
+	// KeepCaller indicates whether the Canvas will
+	// retain caller information for the actions.
+	KeepCaller bool
+}
+
+// Action is a vector graphics action as defined by the
+// vg.Canvas interface. Each method of vg.Canvas must have
+// a corresponding Action type.
+type Action interface {
+	VgCall() string
+	ApplyTo(vg.Canvas)
+	callRecorder() *callRecorder
+}
+
+type callRecorder struct {
+	haveCaller bool
+	file       string
+	line       int
+}
+
+func (c *callRecorder) caller() {
+	_, c.file, c.line, c.haveCaller = runtime.Caller(3)
+}
+
+func (c callRecorder) String() string {
+	if !c.haveCaller {
+		return ""
+	}
+	return fmt.Sprintf("%s:%d ", c.file, c.line)
+}
+
+// New returns a new Canvas pre-filled with the given base actions.
+func New(dpi float64) *Canvas {
+	return &Canvas{
+		Resolution: dpi,
+	}
+}
+
+// Reset resets the Canvas to the base state.
+func (c *Canvas) Reset() {
+	c.Actions = c.Actions[:0]
+}
+
+// ReplayOn applies the set of Actions recorded by the Recorder onto
+// the destination Canvas.
+func (c *Canvas) ReplayOn(dst vg.Canvas) error {
+	fonts := make(map[fontID]vg.Font)
+	for _, a := range c.Actions {
+		if fa, ok := a.(*FillString); ok {
```

Comment:
I prefer to reduce this deep indentation like this:
fa, ok := a.(*FillString)
if !ok {
  continue
}

Up to you.